plutus-core/plutus-ir/src/PlutusIR/Arity.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE GADTs      #-}
+{-# LANGUAGE RankNTypes #-}
+module PlutusIR.Arity where
+
+import Control.Lens
+import PlutusIR.Core.Type
+import Prettyprinter
+
+-- | Is the next argument a term or a type?
+data Param =
+    TermParam | TypeParam
+    deriving stock (Show, Eq)
+
+instance Pretty Param where
+  pretty = viaShow
+
+{-|
+The (syntactic) arity of a term. That is, a record of the arguments that the
+term expects before it may do some work. Since we have both type and lambda
+abstractions, this is not a simple argument count, but rather a list of values
+indicating whether the next argument should be a term or a type.
+
+Note that this is the syntactic arity, i.e. it just corresponds to the number of
+syntactic lambda and type abstractions on the outside of the term. It is thus
+an under-approximation of how many arguments the term may need.
+e.g. consider the term @let id = \x -> x in id@: the variable @id@ has syntactic
+arity @[]@, but does in fact need an argument before it does any work.
+-}
+type Arity = [Param]
```

Comment:
I guess I think about it differently. For me arity is a "summary of how the function is going to be used in the program". As such it makes sense to me to include information about the parameters, because that's what you need!